### PR TITLE
Implement stale job cleanup & graceful exit handling

### PR DESCRIPTION
**Pull request by ali-sharif**
Implements a process which runs every hour to delete any one-shot jobs (of type `normal`) which concluded in the last hour.

### CODE REVIEW

**Comment by dev-dennis-guye** [comment]
~~Q: will a cleanup cross services eg. will a cleanup in service A touch one-off tasks in service B?~~

Yes.

File: bin/utils.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        let totalJobCount = 0;
+
+        const collection = agenda._collection || agenda._collection.collection;
+        if (!collection) {
+          console.log(`Cannot perform cleanup: agenda.collection not found!`);
+          return;
+        }
+
+        const now = dayjs();
+        const expirationTime = now.subtract(JOB_EXPIRATION_DURATION);
```
Comment (qoire) [inline bin/utils.js:62]: Just verifying my assumption, this cron will clean up normal finished tasks that are hour one hour expired

Comment (qoire) [reply bin/utils.js:62]: Nevermind, question answered in PR description, all good

File: bin/utils.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        console.log(`Cleaned-up ${cleanupCount}/${totalJobCount} one-shot job(s)! Next cleanup runs ${CLEANUP_INTERVAL_STR}.`);
+      } catch (e) {
+        console.log(`Stale job cleanup procedure failed with exception: `, { error: serializeError(e) })
+      }
+    })();
+
+    setTimeout(cleanup, CLEANUP_INTERVAL_MS);
+  };
+
+  agenda.on("ready", async () => {
```
Comment (qoire) [inline bin/utils.js:96]: Not critical: would `once` be a better fit here? But I'm pretty sure we have guarantees of `ready` only firing once.

Comment (qoire) [reply bin/utils.js:96]: Eh, can confirm this only gets called once on Agenda object instantiation so... no biggy, can ignore

**Review by qoire — APPROVED** [review]

File: bin/utils.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      exitCallback(0);
+    });
+  });
+  process
+    /* NodeJS process will exit when:
+     * - the event loop's queue is empty.
+     * - no background/asynchronous tasks remain that are capable of adding to the queue. */
+    .on('exit', () => {
+      console.log('NodeJS has nothing more to execute ...');
+      exitCallback();
```
Comment (dev-dennis-guye) [inline bin/utils.js:27]: https://nodejs.org/api/process.html#process_event_exit

From node.js documentation 

> Listener functions must only perform synchronous operations. The Node.js process will exit immediately after calling the 'exit' event listeners causing any additional work still queued in the event loop to be abandoned.

Comment (dev-dennis-guye) [reply bin/utils.js:27]: I think that you want this https://nodejs.org/api/process.html#event-beforeexit

**Pull request merged**